File: python/src/memosift/core/deduplicator.py
# Layer 2: Semantic deduplication — exact hash + MinHash/LSH fuzzy + TF-IDF fallback.
from __future__ import annotations

import hashlib
import math
import re
from collections import Counter
from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from memosift.core.types import (
    ClassifiedMessage,
    CompressionPolicy,
    ContentType,
    DependencyMap,
)

if TYPE_CHECKING:
    from memosift.config import MemoSiftConfig

# Policies that skip deduplication entirely.
_SKIP_DEDUP_POLICIES = {CompressionPolicy.PRESERVE, CompressionPolicy.LIGHT}

# MinHash parameters tuned for 0.85 Jaccard threshold.
_NUM_HASHES = 128
_NUM_BANDS = 16
_ROWS_PER_BAND = _NUM_HASHES // _NUM_BANDS  # 8

# Use TF-IDF for groups smaller than this; MinHash overhead isn't worth it.
# Lowered from 10 to 5 (Item 4.1) — 128 permutations is fast enough for small groups.
_MINHASH_MIN_GROUP_SIZE = 5

# Shingle size for MinHash (character n-grams).
_SHINGLE_SIZE = 5

# Large prime for MinHash hash functions.
_PRIME = 2**61 - 1

# Pre-computed hash function coefficients (deterministic seed).
_HASH_A = [((i + 1) * 6364136223846793005 + 1) % _PRIME for i in range(_NUM_HASHES)]
_HASH_B = [((i + 1) * 1442695040888963407 + 7) % _PRIME for i in range(_NUM_HASHES)]


@dataclass
class CrossWindowState:
    """State shared across compression windows for cross-window dedup.

    Pass the same instance to multiple ``deduplicate()`` calls to catch
    duplicates that span windows.
    """

    content_hashes: set[str] = field(default_factory=set)


def deduplicate(
    segments: list[ClassifiedMessage],
    config: MemoSiftConfig,
    cross_window: CrossWindowState | None = None,
    exact_only: bool = False,
) -> tuple[list[ClassifiedMessage], DependencyMap]:
    """Deduplicate messages via exact hash matching, fuzzy similarity, and chunk-level dedup.

    Uses MinHash/LSH for groups >= 5 messages (lowered from 10 per Item 4.1).
    Falls back to TF-IDF cosine similarity for smaller groups.
    Long messages (>1000 tokens) are split into chunks and deduped at chunk level.
    Messages with PRESERVE or LIGHT policy are never deduplicated.

    Args:
        segments: Classified messages from Layer 1.
        config: Pipeline configuration (controls ``dedup_similarity_threshold``).
        cross_window: Optional state for cross-window dedup.
        exact_only: When True, skip fuzzy and chunk-level dedup (faster).

    Returns:
        A tuple of (deduplicated segments, dependency map for Layer 6).
    """
    deps = DependencyMap()
    result = _exact_dedup(segments, deps, cross_window)
    if not exact_only:
        result = _fuzzy_dedup(result, config.dedup_similarity_threshold, deps)
        result = _chunk_dedup(result, config.dedup_similarity_threshold, deps)
    return result, deps


# ── Exact-match deduplication ───────────────────────────────────────────────


def _normalize_for_hash(text: str) -> str:
    """Normalize text before hashing: collapse whitespace, strip edges."""
    return re.sub(r"\s+", " ", text).strip()


def _content_hash(text: str) -> str:
    """SHA-256 hash of normalized text."""
    return hashlib.sha256(_normalize_for_hash(text).encode("utf-8")).hexdigest()


def _exact_dedup(
    segments: list[ClassifiedMessage],
    deps: DependencyMap,
    cross_window: CrossWindowState | None = None,
) -> list[ClassifiedMessage]:
    """Replace exact-duplicate messages with back-references.

    Messages with PRESERVE or LIGHT policy are skipped.
    The later duplicate is replaced; the earlier one is kept.
    If ``cross_window`` is provided, also checks against hashes from
    previous compression windows.
    """
    seen: dict[str, int] = {}  # hash → index in the *segments* input list
    result: list[ClassifiedMessage] = []

    for i, seg in enumerate(segments):
        if seg.policy in _SKIP_DEDUP_POLICIES:
            result.append(seg)
            continue

        h = _content_hash(seg.content)

        # Check against current window.
        if h in seen:
            original_seg_idx = seen[h]
            original = segments[original_seg_idx]
            tool_name = seg.message.name or "content"
            ref_text = (
                f"[{tool_name} was read earlier in this session. "
                f"Content unchanged. See message #{original.original_index}.]"
            )
            new_seg = _replace_content(seg, ref_text)
            deps.add(seg.original_index, original.original_index)
            result.append(new_seg)
        elif cross_window is not None and h in cross_window.content_hashes:
            # Seen in a previous window — replace with cross-window reference.
            tool_name = seg.message.name or "content"
            ref_text = (
                f"[{tool_name} was read earlier in this session. "
                f"Content unchanged (seen in previous context window).]"
            )
            result.append(_replace_content(seg, ref_text))
        else:
            seen[h] = i
            result.append(seg)

    # Add all hashes from this window to cross-window state.
    if cross_window is not None:
        cross_window.content_hashes.update(seen.keys() for _ in [])  # no-op placeholder
        for h in seen:
            cross_window.content_hashes.add(h)

    return result


# ── Fuzzy deduplication (MinHash/LSH + TF-IDF fallback) ────────────────────


def _fuzzy_dedup(
    segments: list[ClassifiedMessage],
    threshold: float,
    deps: DependencyMap,
) -> list[ClassifiedMessage]:
    """Fuzzy dedup using MinHash/LSH for large groups, TF-IDF for small groups.

    Only compares messages of the same ``ContentType``.
    Messages with PRESERVE or LIGHT policy are skipped.
    """
    groups: dict[ContentType, list[int]] = {}
    for i, seg in enumerate(segments):
        if seg.policy in _SKIP_DEDUP_POLICIES:
            continue
        groups.setdefault(seg.content_type, []).append(i)

    for indices in groups.values():
        if len(indices) < 2:
            continue

        docs = [segments[i].content for i in indices]
        if not any(docs):
            continue

        if len(indices) >= _MINHASH_MIN_GROUP_SIZE:
            _fuzzy_dedup_minhash(segments, indices, docs, threshold, deps)
        else:
            _fuzzy_dedup_tfidf(segments, indices, docs, threshold, deps)

    return segments


def _fuzzy_dedup_minhash(
    segments: list[ClassifiedMessage],
    indices: list[int],
    docs: list[str],
    threshold: float,
    deps: DependencyMap,
) -> None:
    """MinHash/LSH fuzzy dedup — O(n) after preprocessing."""
    # Compute MinHash signatures.
    signatures = [_minhash_signature(doc) for doc in docs]

    # LSH banding: hash each band to find candidate pairs.
    candidates: set[tuple[int, int]] = set()
    for band in range(_NUM_BANDS):
        buckets: dict[int, list[int]] = {}
        start = band * _ROWS_PER_BAND
        end = start + _ROWS_PER_BAND
        for pos, sig in enumerate(signatures):
            band_hash = hash(tuple(sig[start:end]))
            buckets.setdefault(band_hash, []).append(pos)

        for bucket_members in buckets.values():
            if len(bucket_members) < 2:
                continue
            for a in range(len(bucket_members)):
                for b in range(a + 1, len(bucket_members)):
                    candidates.add((bucket_members[a], bucket_members[b]))

    # Verify candidates with full Jaccard similarity.
    deduped: set[int] = set()
    for a_pos, b_pos in sorted(candidates):
        if a_pos in deduped or b_pos in deduped:
            continue

        sim = _jaccard_from_minhash(signatures[a_pos], signatures[b_pos])
        if sim >= threshold:
            _mark_as_dedup(segments, indices, a_pos, b_pos, deps)
            deduped.add(a_pos)


def _fuzzy_dedup_tfidf(
    segments: list[ClassifiedMessage],
    indices: list[int],
    docs: list[str],
    threshold: float,
    deps: DependencyMap,
) -> None:
    """TF-IDF cosine similarity fuzzy dedup — fallback for small groups."""
    vectors = _tfidf_vectors(docs)
    deduped: set[int] = set()

    for a_pos in range(len(indices)):
        if a_pos in deduped:
            continue
        for b_pos in range(a_pos + 1, len(indices)):
            if b_pos in deduped:
                continue
            sim = _cosine_similarity(vectors[a_pos], vectors[b_pos])
            if sim >= threshold:
                _mark_as_dedup(segments, indices, a_pos, b_pos, deps)
                deduped.add(a_pos)
                break


def _mark_as_dedup(
    segments: list[ClassifiedMessage],
    indices: list[int],
    a_pos: int,
    b_pos: int,
    deps: DependencyMap,
) -> None:
    """Mark the older message (a_pos) as a duplicate of the newer (b_pos)."""
    older_idx = indices[a_pos]
    newer_idx = indices[b_pos]
    older_seg = segments[older_idx]
    newer_seg = segments[newer_idx]
    tool_name = older_seg.message.name or "content"
    ref_text = (
        f"[{tool_name} was read earlier in this session. "
        f"Content unchanged. See message #{newer_seg.original_index}.]"
    )
    segments[older_idx] = _replace_content(older_seg, ref_text)
    deps.add(older_seg.original_index, newer_seg.original_index)


# ── MinHash implementation (zero dependencies) ────────────────────────────


def _shingles(text: str, k: int = _SHINGLE_SIZE) -> set[int]:
    """Generate character k-gram shingle hashes from text."""
    text = text.lower().strip()
    if len(text) < k:
        return {hash(text)} if text else set()
    return {hash(text[i : i + k]) for i in range(len(text) - k + 1)}


def _minhash_signature(text: str) -> list[int]:
    """Compute MinHash signature (list of N minimum hash values)."""
    shingle_set = _shingles(text)
    if not shingle_set:
        return [_PRIME] * _NUM_HASHES

    sig = [_PRIME] * _NUM_HASHES
    for shingle in shingle_set:
        for i in range(_NUM_HASHES):
            h = (_HASH_A[i] * shingle + _HASH_B[i]) % _PRIME
            if h < sig[i]:
                sig[i] = h
    return sig


def _jaccard_from_minhash(sig_a: list[int], sig_b: list[int]) -> float:
    """Estimate Jaccard similarity from two MinHash signatures."""
    if not sig_a or not sig_b:
        return 0.0
    matches = sum(1 for a, b in zip(sig_a, sig_b, strict=False) if a == b)
    return matches / len(sig_a)


# ── TF-IDF implementation (zero dependencies) ──────────────────────────────


def _tokenize(text: str) -> list[str]:
    """Split text into lowercase word tokens."""
    return re.findall(r"\b\w+\b", text.lower())


def _tfidf_vectors(documents: list[str]) -> list[dict[str, float]]:
    """Compute sparse TF-IDF vectors for a list of documents.

    IDF formula: log((N + 1) / (df + 1)) + 1  (scikit-learn default).
    """
    n_docs = len(documents)
    tokenized = [_tokenize(doc) for doc in documents]

    doc_freq: Counter[str] = Counter()
    for tokens in tokenized:
        doc_freq.update(set(tokens))

    idf: dict[str, float] = {
        token: math.log((n_docs + 1) / (freq + 1)) + 1 for token, freq in doc_freq.items()
    }

    vectors: list[dict[str, float]] = []
    for tokens in tokenized:
        if not tokens:
            vectors.append({})
            continue
        tf: Counter[str] = Counter(tokens)
        total = len(tokens)
        vec = {token: (count / total) * idf.get(token, 1.0) for token, count in tf.items()}
        vectors.append(vec)

    return vectors


def _cosine_similarity(a: dict[str, float], b: dict[str, float]) -> float:
    """Cosine similarity between two sparse vectors."""
    if not a or not b:
        return 0.0

    shared_keys = a.keys() & b.keys()
    dot = sum(a[k] * b[k] for k in shared_keys)

    norm_a = math.sqrt(sum(v * v for v in a.values()))
    norm_b = math.sqrt(sum(v * v for v in b.values()))

    if norm_a == 0 or norm_b == 0:
        return 0.0

    return dot / (norm_a * norm_b)


# ── Chunk-level deduplication ──────────────────────────────────────────────

# Minimum estimated tokens for a message to be eligible for chunk dedup.
_CHUNK_DEDUP_MIN_TOKENS = 1000

# Target chunk size in tokens (approximate).
_CHUNK_TARGET_TOKENS = 200

# Similarity threshold for chunk dedup (slightly higher than message-level).
_CHUNK_SIMILARITY_THRESHOLD = 0.85


def _chunk_dedup(
    segments: list[ClassifiedMessage],
    threshold: float,
    deps: DependencyMap,
) -> list[ClassifiedMessage]:
    """Chunk-level dedup for long messages.

    Splits long messages (>1000 tokens) into ~200-token chunks at paragraph
    boundaries (prose) or function/class boundaries (code). Deduplicates
    chunks across segments — if a chunk is >85% similar to a previously seen
    chunk, replace with back-reference keeping only the diff.

    Targets tool results — the biggest token consumers.
    """
    seen_chunks: dict[str, tuple[int, str]] = {}  # hash → (seg_index, first_line_label)
    result: list[ClassifiedMessage] = []

    for seg in segments:
        if seg.policy in _SKIP_DEDUP_POLICIES:
            result.append(seg)
            continue

        # Only chunk-dedup long messages.
        est_tokens = len((seg.content or "").split())
        if est_tokens < _CHUNK_DEDUP_MIN_TOKENS:
            result.append(seg)
            continue

        # Split into chunks.
        if seg.content_type in {ContentType.CODE_BLOCK}:
            chunks = _split_code_chunks(seg.content or "")
        else:
            chunks = _split_paragraph_chunks(seg.content or "")

        if len(chunks) <= 1:
            result.append(seg)
            continue

        # Check each chunk for duplicates.
        new_chunks: list[str] = []
        any_deduped = False
        for chunk in chunks:
            chunk_hash = hashlib.sha256(
                re.sub(r"\s+", " ", chunk).strip().encode("utf-8")
            ).hexdigest()

            if chunk_hash in seen_chunks:
                orig_idx, label = seen_chunks[chunk_hash]
                new_chunks.append(f"[Duplicate chunk — see message {orig_idx}: {label}]")
                any_deduped = True
            else:
                seen_chunks[chunk_hash] = (seg.original_index, chunk[:60].strip())
                new_chunks.append(chunk)

        if any_deduped:
            new_content = "\n\n".join(new_chunks)
            result.append(_replace_content(seg, new_content))
        else:
            result.append(seg)

    return result


def _split_paragraph_chunks(text: str) -> list[str]:
    """Split text into chunks at paragraph boundaries (~200 tokens each)."""
    paragraphs = re.split(r"\n\s*\n", text)
    chunks: list[str] = []
    current: list[str] = []
    current_tokens = 0

    for para in paragraphs:
        para_tokens = len(para.split())
        if current_tokens + para_tokens > _CHUNK_TARGET_TOKENS and current:
            chunks.append("\n\n".join(current))
            current = [para]
            current_tokens = para_tokens
        else:
            current.append(para)
            current_tokens += para_tokens

    if current:
        chunks.append("\n\n".join(current))

    return chunks


def _split_code_chunks(text: str) -> list[str]:
    """Split code into chunks at function/class boundaries.

    Uses regex to detect function and class definitions. Falls back to
    paragraph splitting if no code boundaries are found.
    """
    # Try to split at function/class boundaries.
    boundary_re = re.compile(
        r"^(?=\s*(?:def |class |function |async |export ))",
        re.MULTILINE,
    )
    parts = boundary_re.split(text)
    parts = [p for p in parts if p.strip()]

    if len(parts) <= 1:
        return _split_paragraph_chunks(text)

    # Group small parts together to meet minimum chunk size.
    chunks: list[str] = []
    current: list[str] = []
    current_tokens = 0
    for part in parts:
        part_tokens = len(part.split())
        if current_tokens + part_tokens > _CHUNK_TARGET_TOKENS and current:
            chunks.append("\n".join(current))
            current = [part]
            current_tokens = part_tokens
        else:
            current.append(part)
            current_tokens += part_tokens

    if current:
        chunks.append("\n".join(current))

    return chunks


# ── Helpers ─────────────────────────────────────────────────────────────────


def _replace_content(seg: ClassifiedMessage, new_content: str) -> ClassifiedMessage:
    """Create a new ClassifiedMessage with replaced content (immutable pattern)."""
    from dataclasses import replace as dc_replace

    from memosift.core.types import MemoSiftMessage

    new_msg = MemoSiftMessage(
        role=seg.message.role,
        content=new_content,
        name=seg.message.name,
        tool_call_id=seg.message.tool_call_id,
        tool_calls=seg.message.tool_calls,
        metadata=seg.message.metadata,
    )
    return dc_replace(seg, message=new_msg)

File: typescript/src/core/deduplicator.ts
// Layer 2: Semantic deduplication — exact hash + MinHash/LSH fuzzy + TF-IDF fallback.

import { createHash } from "node:crypto";
import type { MemoSiftConfig } from "../config.js";
import {
  type ClassifiedMessage,
  CompressionPolicy,
  ContentType,
  type CrossWindowState,
  type DependencyMap,
  type MemoSiftMessage,
  createClassified,
  createDependencyMap,
  createMessage,
  depMapAdd,
} from "./types.js";

// Policies that skip deduplication entirely.
const SKIP_DEDUP_POLICIES = new Set([CompressionPolicy.PRESERVE, CompressionPolicy.LIGHT]);

// MinHash parameters tuned for 0.85 Jaccard threshold.
const NUM_HASHES = 128;
const NUM_BANDS = 16;
const ROWS_PER_BAND = NUM_HASHES / NUM_BANDS; // 8

// Use TF-IDF for groups smaller than this; MinHash overhead isn't worth it.
const MINHASH_MIN_GROUP_SIZE = 5;

// Shingle size for MinHash (character n-grams).
const SHINGLE_SIZE = 5;

// 32-bit safe prime for MinHash hash functions.
const PRIME = 2147483647; // 2^31 - 1

// Large primes for hash coefficient generation.
const LARGE_PRIME_A = 1610612741;
const LARGE_PRIME_B = 805306457;

// Pre-computed hash function coefficients (deterministic seed).
const HASH_A: number[] = [];
const HASH_B: number[] = [];
for (let i = 0; i < NUM_HASHES; i++) {
  HASH_A.push((((i + 1) * LARGE_PRIME_A + 1) % PRIME) >>> 0);
  HASH_B.push((((i + 1) * LARGE_PRIME_B + 7) % PRIME) >>> 0);
}

export function deduplicate(
  segments: ClassifiedMessage[],
  config: MemoSiftConfig,
  crossWindow?: CrossWindowState | null,
  exactOnly?: boolean,
): { segments: ClassifiedMessage[]; deps: DependencyMap } {
  const deps: DependencyMap = createDependencyMap();
  let result = exactDedup(segments, deps, crossWindow);
  if (!exactOnly) {
    result = fuzzyDedup(result, config.dedupSimilarityThreshold, deps);
    result = chunkDedup(result, deps);
  }
  return { segments: result, deps };
}

function normalizeForHash(text: string): string {
  return text.replace(/\s+/g, " ").trim();
}

function contentHash(text: string): string {
  return createHash("sha256").update(normalizeForHash(text)).digest("hex");
}

function exactDedup(
  segments: ClassifiedMessage[],
  deps: DependencyMap,
  crossWindow?: CrossWindowState | null,
): ClassifiedMessage[] {
  const seen = new Map<string, number>(); // hash → index in input segments list
  const result: ClassifiedMessage[] = [];

  for (let i = 0; i < segments.length; i++) {
    const seg = segments[i]!;
    if (SKIP_DEDUP_POLICIES.has(seg.policy)) {
      result.push(seg);
      continue;
    }
    const h = contentHash(seg.message.content);
    if (seen.has(h)) {
      const originalSegIdx = seen.get(h)!;
      const original = segments[originalSegIdx]!;
      const toolName = seg.message.name ?? "content";
      const refText = `[${toolName} was read earlier in this session. Content unchanged. See message #${original.originalIndex}.]`;
      const newMsg = createMessage(seg.message.role, refText, {
        name: seg.message.name,
        toolCallId: seg.message.toolCallId,
        toolCalls: seg.message.toolCalls,
        metadata: seg.message.metadata,
      });
      depMapAdd(deps, seg.originalIndex, original.originalIndex);
      result.push(
        createClassified(newMsg, seg.contentType, seg.policy, {
          originalIndex: seg.originalIndex,
          relevanceScore: seg.relevanceScore,
          protected: seg.protected,
        }),
      );
    } else if (crossWindow?.contentHashes.has(h)) {
      // Seen in a previous window.
      const toolName = seg.message.name ?? "content";
      const refText = `[${toolName} was read earlier in this session. Content unchanged (seen in previous context window).]`;
      const newMsg = createMessage(seg.message.role, refText, {
        name: seg.message.name,
        toolCallId: seg.message.toolCallId,
        toolCalls: seg.message.toolCalls,
        metadata: seg.message.metadata,
      });
      result.push(
        createClassified(newMsg, seg.contentType, seg.policy, {
          originalIndex: seg.originalIndex,
          relevanceScore: seg.relevanceScore,
          protected: seg.protected,
        }),
      );
    } else {
      seen.set(h, i);
      result.push(seg);
    }
  }

  // Add all hashes to cross-window state.
  if (crossWindow) {
    for (const h of seen.keys()) {
      crossWindow.contentHashes.add(h);
    }
  }

  return result;
}

function fuzzyDedup(
  segments: ClassifiedMessage[],
  threshold: number,
  deps: DependencyMap,
): ClassifiedMessage[] {
  const groups = new Map<string, number[]>();
  for (let i = 0; i < segments.length; i++) {
    const seg = segments[i]!;
    if (SKIP_DEDUP_POLICIES.has(seg.policy)) continue;
    const key = seg.contentType;
    if (!groups.has(key)) groups.set(key, []);
    groups.get(key)!.push(i);
  }

  for (const indices of groups.values()) {
    if (indices.length < 2) continue;
    const docs = indices.map((i) => segments[i]!.message.content);
    if (!docs.some((d) => d)) continue;

    if (indices.length >= MINHASH_MIN_GROUP_SIZE) {
      fuzzyDedupMinhash(segments, indices, docs, threshold, deps);
    } else {
      fuzzyDedupTfidf(segments, indices, docs, threshold, deps);
    }
  }

  return segments;
}

/** Mark older message (aPos) as duplicate of newer (bPos). Mutates segments array. */
function markAsDedup(
  segments: ClassifiedMessage[],
  indices: number[],
  aPos: number,
  bPos: number,
  deps: DependencyMap,
): void {
  const olderIdx = indices[aPos]!;
  const newerIdx = indices[bPos]!;
  const older = segments[olderIdx]!;
  const newer = segments[newerIdx]!;
  const toolName = older.message.name ?? "content";
  const refText = `[${toolName} was read earlier in this session. Content unchanged. See message #${newer.originalIndex}.]`;
  const newMsg = createMessage(older.message.role, refText, {
    name: older.message.name,
    toolCallId: older.message.toolCallId,
    toolCalls: older.message.toolCalls,
    metadata: older.message.metadata,
  });
  segments[olderIdx] = createClassified(newMsg, older.contentType, older.policy, {
    originalIndex: older.originalIndex,
    relevanceScore: older.relevanceScore,
    protected: older.protected,
  });
  depMapAdd(deps, older.originalIndex, newer.originalIndex);
}

/** MinHash/LSH fuzzy dedup — O(n) after preprocessing. */
function fuzzyDedupMinhash(
  segments: ClassifiedMessage[],
  indices: number[],
  docs: string[],
  threshold: number,
  deps: DependencyMap,
): void {
  // Compute MinHash signatures.
  const signatures = docs.map((doc) => minhashSignature(doc));

  // LSH banding: hash each band to find candidate pairs.
  const candidates = new Set<string>(); // "aPos,bPos" encoded as string
  for (let band = 0; band < NUM_BANDS; band++) {
    const buckets = new Map<string, number[]>();
    const start = band * ROWS_PER_BAND;
    const end = start + ROWS_PER_BAND;

    for (let pos = 0; pos < signatures.length; pos++) {
      const sig = signatures[pos]!;
      // Create a deterministic band hash from the signature slice.
      const bandKey = sig.slice(start, end).join(",");
      if (!buckets.has(bandKey)) buckets.set(bandKey, []);
      buckets.get(bandKey)!.push(pos);
    }

    for (const members of buckets.values()) {
      if (members.length < 2) continue;
      for (let a = 0; a < members.length; a++) {
        for (let b = a + 1; b < members.length; b++) {
          const pa = members[a]!;
          const pb = members[b]!;
          // Always store smaller index first for dedup consistency.
          const key = pa < pb ? `${pa},${pb}` : `${pb},${pa}`;
          candidates.add(key);
        }
      }
    }
  }

  // Verify candidates with full Jaccard similarity.
  const deduped = new Set<number>();
  // Sort candidates for deterministic processing order.
  const sortedCandidates = [...candidates].sort();

  for (const key of sortedCandidates) {
    const [aStr, bStr] = key.split(",");
    const aPos = Number(aStr);
    const bPos = Number(bStr);

    if (deduped.has(aPos) || deduped.has(bPos)) continue;

    const sim = jaccardFromMinhash(signatures[aPos]!, signatures[bPos]!);
    if (sim >= threshold) {
      markAsDedup(segments, indices, aPos, bPos, deps);
      deduped.add(aPos);
    }
  }
}

/** TF-IDF cosine similarity fuzzy dedup — fallback for small groups. */
function fuzzyDedupTfidf(
  segments: ClassifiedMessage[],
  indices: number[],
  docs: string[],
  threshold: number,
  deps: DependencyMap,
): void {
  const vectors = tfidfVectors(docs);
  const deduped = new Set<number>();

  for (let aPos = 0; aPos < indices.length; aPos++) {
    if (deduped.has(aPos)) continue;
    for (let bPos = aPos + 1; bPos < indices.length; bPos++) {
      if (deduped.has(bPos)) continue;
      const sim = cosineSimilarity(vectors[aPos]!, vectors[bPos]!);
      if (sim >= threshold) {
        markAsDedup(segments, indices, aPos, bPos, deps);
        deduped.add(aPos);
        break;
      }
    }
  }
}

// ── MinHash implementation (zero dependencies) ──────────────────────────────

/** Simple 32-bit string hash (FNV-1a variant). */
function hashString32(s: string): number {
  let hash = 0x811c9dc5; // FNV offset basis
  for (let i = 0; i < s.length; i++) {
    hash ^= s.charCodeAt(i);
    // FNV-1a multiply: multiply by FNV prime 16777619, keep 32-bit.
    hash = Math.imul(hash, 16777619);
  }
  return hash >>> 0; // ensure unsigned 32-bit
}

/** Generate character k-gram shingle hashes from text. */
export function shingles(text: string, k: number = SHINGLE_SIZE): Set<number> {
  const t = text.toLowerCase().trim();
  if (t.length < k) {
    return t ? new Set([hashString32(t)]) : new Set();
  }
  const result = new Set<number>();
  for (let i = 0; i <= t.length - k; i++) {
    result.add(hashString32(t.slice(i, i + k)));
  }
  return result;
}

/** Compute MinHash signature (array of NUM_HASHES minimum hash values). */
export function minhashSignature(text: string): number[] {
  const shingleSet = shingles(text);
  if (shingleSet.size === 0) {
    return new Array(NUM_HASHES).fill(PRIME);
  }

  const sig = new Array<number>(NUM_HASHES).fill(PRIME);
  for (const shingle of shingleSet) {
    for (let i = 0; i < NUM_HASHES; i++) {
      // (a * shingle + b) mod PRIME, using 32-bit safe Math.imul + modulo.
      // Since PRIME = 2^31-1, intermediate results fit in safe integer range
      // when we compute step by step.
      const ax = Math.imul(HASH_A[i]!, shingle >>> 0);
      const h = (((ax >>> 0) + HASH_B[i]!) % PRIME) >>> 0;
      if (h < sig[i]!) {
        sig[i] = h;
      }
    }
  }
  return sig;
}

/** Estimate Jaccard similarity from two MinHash signatures. */
export function jaccardFromMinhash(sigA: number[], sigB: number[]): number {
  if (sigA.length === 0 || sigB.length === 0) return 0;
  let matches = 0;
  for (let i = 0; i < sigA.length; i++) {
    if (sigA[i] === sigB[i]) matches++;
  }
  return matches / sigA.length;
}

// ── TF-IDF implementation ──────────────────────────────────────────────────

function tokenize(text: string): string[] {
  return text.toLowerCase().match(/\b\w+\b/g) ?? [];
}

function tfidfVectors(documents: string[]): Map<string, number>[] {
  const nDocs = documents.length;
  const tokenized = documents.map(tokenize);
  const docFreq = new Map<string, number>();

  for (const tokens of tokenized) {
    const unique = new Set(tokens);
    for (const token of unique) {
      docFreq.set(token, (docFreq.get(token) ?? 0) + 1);
    }
  }

  const idf = new Map<string, number>();
  for (const [token, freq] of docFreq) {
    idf.set(token, Math.log((nDocs + 1) / (freq + 1)) + 1);
  }

  return tokenized.map((tokens) => {
    if (tokens.length === 0) return new Map();
    const tf = new Map<string, number>();
    for (const t of tokens) tf.set(t, (tf.get(t) ?? 0) + 1);
    const vec = new Map<string, number>();
    for (const [token, count] of tf) {
      vec.set(token, (count / tokens.length) * (idf.get(token) ?? 1));
    }
    return vec;
  });
}

function cosineSimilarity(a: Map<string, number>, b: Map<string, number>): number {
  if (a.size === 0 || b.size === 0) return 0;
  let dot = 0;
  for (const [k, v] of a) {
    const bv = b.get(k);
    if (bv !== undefined) dot += v * bv;
  }
  let normA = 0;
  for (const v of a.values()) normA += v * v;
  let normB = 0;
  for (const v of b.values()) normB += v * v;
  normA = Math.sqrt(normA);
  normB = Math.sqrt(normB);
  if (normA === 0 || normB === 0) return 0;
  return dot / (normA * normB);
}

// ── Chunk-level deduplication ──────────────────────────────────────────────

const CHUNK_DEDUP_MIN_TOKENS = 1000;
const CHUNK_TARGET_TOKENS = 200;

function chunkDedup(segments: ClassifiedMessage[], deps: DependencyMap): ClassifiedMessage[] {
  const seenChunks = new Map<string, [number, string]>();
  const result: ClassifiedMessage[] = [];

  for (const seg of segments) {
    if (SKIP_DEDUP_POLICIES.has(seg.policy)) {
      result.push(seg);
      continue;
    }

    const estTokens = (seg.message.content || "").split(/\s+/).length;
    if (estTokens < CHUNK_DEDUP_MIN_TOKENS) {
      result.push(seg);
      continue;
    }

    const chunks =
      seg.contentType === ContentType.CODE_BLOCK
        ? splitCodeChunks(seg.message.content)
        : splitParagraphChunks(seg.message.content);

    if (chunks.length <= 1) {
      result.push(seg);
      continue;
    }

    const newChunks: string[] = [];
    let anyDeduped = false;
    for (const chunk of chunks) {
      const chunkHash = createHash("sha256")
        .update(chunk.replace(/\s+/g, " ").trim())
        .digest("hex");

      if (seenChunks.has(chunkHash)) {
        const [origIdx, label] = seenChunks.get(chunkHash)!;
        newChunks.push(`[Duplicate chunk — see message ${origIdx}: ${label}]`);
        anyDeduped = true;
      } else {
        seenChunks.set(chunkHash, [seg.originalIndex, chunk.slice(0, 60).trim()]);
        newChunks.push(chunk);
      }
    }

    if (anyDeduped) {
      const newMsg = createMessage(seg.message.role, newChunks.join("\n\n"), {
        name: seg.message.name,
        toolCallId: seg.message.toolCallId,
        toolCalls: seg.message.toolCalls,
        metadata: seg.message.metadata,
      });
      result.push(
        createClassified(newMsg, seg.contentType, seg.policy, {
          originalIndex: seg.originalIndex,
          relevanceScore: seg.relevanceScore,
          protected: seg.protected,
          importanceScore: seg.importanceScore,
          shield: seg.shield,
        }),
      );
    } else {
      result.push(seg);
    }
  }

  return result;
}

function splitParagraphChunks(text: string): string[] {
  const paragraphs = text.split(/\n\s*\n/);
  const chunks: string[] = [];
  let current: string[] = [];
  let currentTokens = 0;

  for (const para of paragraphs) {
    const paraTokens = para.split(/\s+/).length;
    if (currentTokens + paraTokens > CHUNK_TARGET_TOKENS && current.length > 0) {
      chunks.push(current.join("\n\n"));
      current = [para];
      currentTokens = paraTokens;
    } else {
      current.push(para);
      currentTokens += paraTokens;
    }
  }
  if (current.length > 0) chunks.push(current.join("\n\n"));
  return chunks;
}

function splitCodeChunks(text: string): string[] {
  const parts = text
    .split(/^(?=\s*(?:def |class |function |async |export ))/m)
    .filter((p) => p.trim());
  if (parts.length <= 1) return splitParagraphChunks(text);

  const chunks: string[] = [];
  let current: string[] = [];
  let currentTokens = 0;
  for (const part of parts) {
    const partTokens = part.split(/\s+/).length;
    if (currentTokens + partTokens > CHUNK_TARGET_TOKENS && current.length > 0) {
      chunks.push(current.join("\n"));
      current = [part];
      currentTokens = partTokens;
    } else {
      current.push(part);
      currentTokens += partTokens;
    }
  }
  if (current.length > 0) chunks.push(current.join("\n"));
  return chunks;
}
